### ISSUE TITLE
Avatar Delete Is not Working

### ISSUE THREAD

**Issue by mkinnan** [signalpoint/avatar#5]
Testing on compiled Android app.

When I click **Delete**, it prompts, I click **OK**, then it does the loading wheel and then nothing happens.

**Comment by signalpoint** [comment]
@mkinnan It should reload the current page when it calls `avatar_pageshow` directly in the `success`: https://github.com/signalpoint/avatar/blob/7.x-1.x/avatar.js#L190

Alternatively, I recommend implementing the avatar hook so you can take action when the picture is deleted: https://github.com/signalpoint/avatar#hooks

**Comment by mkinnan** [comment]
I have the avatar hook function in my module. 

Additionally, the avatar doesn't get deleted. Even though the wheel spins and then nothing happens, the avatar is still there.

**Comment by signalpoint** [comment]
@mkinnan console.log() messages, drupal recent log messages, etc....?

**Comment by mkinnan** [comment]
I don't see any error messages anywhere. I need to try USB debugging on compiled app next.

**Comment by mkinnan** [comment]
No error messages on debugging, but here are the messages when the user clicks the delete button:

```
I/chromium(18713): [INFO:CONSOLE(1499)] "DELETE: http://example.com/?q=drupalgap/file/31.json", source: file:///android_asset/www/jdrupal.js (1499)
I/chromium(18713): [INFO:CONSOLE(1509)] "200 - OK", source: file:///android_asset/www/jdrupal.js (1509)
I/chromium(18713): [INFO:CONSOLE(1499)] "PUT: http://example.com/?q=drupalgap/user/217.json", source: file:///android_asset/www/jdrupal.js (1499)
I/chromium(18713): [INFO:CONSOLE(1509)] "200 - OK", source: file:///android_asset/www/jdrupal.js (1509)
```

**Comment by signalpoint** [comment]
@mkinnan Bummer, a 200 OK, means everything is OK of course... so I am stumped on this one. Maybe via a terminal, go to the actual files directory and then do an `ls -la` and see what the owner/group read/write permissions are on the file. What are those values?

**Comment by mkinnan** [comment]
My .../sites/default/files/pictures/ folder has **drwxrwxr-x** permissions.

The files in my **pictures** folder have **-rw-rw-r--** permissions.

**Comment by signalpoint** [comment]
@mkinnan What's the owner and group and the file in question?

**Comment by mkinnan** [comment]
I'll have to check the owner/group when i get back.

If the user can change/add/delete their avatar via the website interface on .../user/edit, then this issue shouldn't be related to permission/owner/group of the file correct?

**Comment by signalpoint** [comment]
@mkinnan Yes, that's a safe assumption. You could also use something like FireFox Poster to mimic the file_delete call. Then if that doesn't work, it's safer to say it's something with your Drupal site causing the issue.

**Comment by mkinnan** [comment]
@signalpoint The owner/group is apache/apache. I'm not sure I follow how to do the delete call ... is it just pasting a URL? Also, should the **delete avatar** ability work in Ripple Emulator?

**Comment by mkinnan** [comment]
@signalpoint ... I did some debugging. I noticed that user's cannot save avatars either. I then noticed **hook_avatar_action()** was not being called. See this snippet of code:

```
function avatar_form_submit(form, form_state) {
  try {
    if (form_state.values.mode == 'delete') {
      drupalgap_confirm(t('Delete this picture?'), {
        confirmCallback: function(button) {
          if (button == 1) { // Ok
            file_delete(form_state.values.fid, {
              success: function(result) {
                console.log(result);
                user_save({
                  uid: form_state.values.uid,
                  picture: 0
                }, {
                  success: function() {
                    console.log('Trigger 1');
                    console.log(result[0]);
                    if (result[0]) {
                      console.log('Trigger 2');
                      avatar_pageshow(form_state.values.uid);
                      module_invoke_all('avatar_action', 'delete', form_state.values);
                    }
                  }
                });
              }
            });
          }
          else if (button == 2) { } // Cancel
        }
      });
      return;
    }
```

**Trigger 1** get's displayed.

**Trigger 2** does not get displayed.

console.log(result) gives:
![user-object](https://cloud.githubusercontent.com/assets/13283396/18151217/b3416756-6faa-11e6-842f-1b76ae3cd55a.PNG)

console.log(result[0]) yields `undefined`

**Comment by mkinnan** [comment]
Any thoughts on what might be going on?

**Comment by mkinnan** [comment]
I've tried giving _777_ permissions to my **files** folder again.

I've tried being logged in as User 1.

I've tried doing just:
`file_delete(form_state.values.fid, {resource:"delete", service:"file"})`

and that doesn't work. It doesn't delete the file. Is that correct for using **file_delete**?

**Comment by mkinnan** [comment]
I have checked console.log, drupal logs, and apache logs with no observation of an error. All boxes are checked for services on the website. All permissions are checked for users.

Has anyone been able to get the module to work in Ripple Emulator and/or compiled app? This would help me figure out if it is module related or my website/configuration.

**Comment by signalpoint** [comment]
@mkinnan Next I'd recommend placing a watchdog entry each step of the way in the Services file delete function: http://cgit.drupalcode.org/services/tree/resources/file_resource.inc?h=7.x-3.x#n283

e.g.

```
watchdog('mkinnan', 'ABC');
```

Also, in your Services endpoint config, you can turn on debugging via a checkbox, then there will be a bunch of watchdog entries about the service call, and you can see what is coming in.

I'm really stumped by this one... a safe test would be, can you delete nodes from the app? If not, then that's more indicative that your website host may be rejecting any DELETE methods to the site.

**Comment by mkinnan** [comment]
@signalpoint ... I can create, update, and **delete** nodes and comments via the app with no problems.

I enabled Services debugging. When I go to delete the avatar, the **Controller** gives the following. 
Should `[name] => cid` be `[name] => fid`?

![services](https://cloud.githubusercontent.com/assets/13283396/18233406/e3a6fa40-72a3-11e6-9a58-d9d14e6f17db.PNG)

**Comment by mkinnan** [comment]
I believe the problem might be Services related. Look at the **DELETE** query that executes when trying to delete the user's avatar that has been added via the website GUI.

DELETE query when attempting to delete via app and UPDATE query right after that I don't understand its purpose because that file is supposedly already gone via the DELETE query.

```
 // affects 0 rows because there isn't a row that meets those conditions
 DELETE FROM file_usage WHERE (module = 'services') AND (fid = '5') AND (count <= '1') 
 // affects 0 rows
 UPDATE file_usage SET count=count - '1'    WHERE (module = 'services') AND (fid = '5') 
```

delete avatar via website ... 

```
DELETE FROM file_usage WHERE (module = 'user') AND (fid = '5') AND (type = 'user') AND (id = '1') AND (count <= '1') 
UPDATE users SET picture='0' WHERE (picture = '5') 
DELETE FROM file_managed WHERE (fid = '5') 
DELETE FROM file_usage WHERE (fid = '5') 
```

In the `file_usage` table, the **module** value for the user's avatar image is _user_ and the **type** is _user_. Thus, the DELETE query that executes from the app will never delete the user's avatar.

**Comment by mkinnan** [comment]
In the function you linked to above:
http://cgit.drupalcode.org/services/tree/resources/file_resource.inc?h=7.x-3.x#n283

As more confirmation to the above, change **services** to **user** and the app deletes the avatar in Ripple Emulator and compiled Android with no issues. Here's what works:

```
function _file_resource_delete($fid) {
  if ($file = file_load($fid)) {
    file_usage_delete($file, 'user');
    return file_delete($file);
  }
  return FALSE;
}
```

I'm also assuming the reason users cannot add avatars is due to the same type of issue.

**Comment by signalpoint** [comment]
@mkinnan Thank you very much for the thorough debugging here, you definitely uncovered the issue.

First, I've updated the `cid` to `fid`, although that won't actually fix it, see here: https://www.drupal.org/node/2795813

Regarding deleting from the website vs. deleting from the app... that's a very interesting find. As it stands, the Services module can only delete files that it creates. I've filed an issue with the Services module here: https://www.drupal.org/node/2795863

So theoretically you should be able to add/delete profile pictures using only the app. 

It's a complex issue, so I'll wait to hear from other Services maintainers before trying to make any head way on this.

**Comment by mkinnan** [comment]
@signalpoint Glad I could help. It seems this isn't going to be an easy fix ... or resolved in the near term.

Here's a hack I would try out to get things rolling for me but I don't know if this would be the right approach:

```
function _file_resource_delete($fid) {

  $query_module = // do query to get module value from file_usage

  $query_type = // do query to get type value from file_usage

  if ($query_module == $query_type == 'user') {
    if ($file = file_load($fid)) {
     file_usage_delete($file, 'user');
      return file_delete($file);
    }
  } else {
    if ($file = file_load($fid)) {
     file_usage_delete($file, 'services');
      return file_delete($file);
    }
  }

  return FALSE;

}
```

**Comment by signalpoint** [comment]
@mkinnan That hack/patch idea you have would work fine. I'll look to get that issue fixed in the Services module, because I'm sure I'll need it to sometime soon.

In the mean time, I'm 99% certain things should work as expected if the user is adding/editing/removing their avatar from only within the application. Does that work for you?

**Comment by mkinnan** [comment]
@signalpoint So, adding avatars via the app doesn't work either for me. Once I click Save Photo, that button goes away and then nothing else happens. I did try some debugging a week ago and couldn't get any functions to execute to show console log messages. I'll create another issue ticket.
